sssd_test_framework/utils/dbus_types.py

```diff
@@ -0,0 +1,675 @@
+"""D-Bus Type mapping to and from Python Types"""
+
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+
+__all__ = [
+    "DBUSResult",
+    "DBUSType",
+    "DBUSTypeBoolean",
+    "DBUSTypeString",
+    "DBUSTypeObjectPath",
+    "DBUSTypeInteger",
+    "DBUSTypeByte",
+    "DBUSTypeInt16",
+    "DBUSTypeInt32",
+    "DBUSTypeInt64",
+    "DBUSTypeUInt16",
+    "DBUSTypeUInt32",
+    "DBUSTypeUInt64",
+    "DBUSTypeDouble",
+    "DBUSTypeContainer",
+    "DBUSTypeArray",
+    "DBUSTypeDict",
+    "DBUSTypeVariant",
+    "DBUSSignatureReader",
+]
+
+
+# A NOTE ABOUT DBUS-SEND
+# **********************
+# It is important to know that dbus-send does not have a symmetric
+# behavior with relation to how the types are described in text
+# form when used as parameters or results.
+
+
+# TYPES PENDING IMPLEMENTATION
+# ****************************
+# The following D-Bus types are not implemented because they are not
+# currently recognized by dbus-send:
+# - Signature
+# - UNIX FD
+# - Struct
+
+
+class MutableString:
+    """A helper class implementing an extremely simple mutable string"""
+
+    def __init__(self, text: str):
+        self.text = text
+
+    def __str__(self):
+        return self.text
+
+
+class DBUSResult(MutableString):
+    """
+    The result from the D-Bus operation executed by dbus-send.
+    It is used to simplify parsing of the result as the string
+    must be modified and have no carriage return.
+    """
+
+    def __init__(self, text: str):
+        self.text = text.replace("\n", "")
+
+
+class DBUSType(ABC):
+    """
+    Abstract class to create classes that handle type conversion
+    between D-Bus and Python.
+    """
+
+    # It would be good to define here the following abstract class members:
+    #    _type_prefix
+    #
+    # But there is not such thing as _abstract class members_ in Python and
+    # linters will complain although it will work because the contrete
+    # subclasses will declare the concrete member. That's why we are ignoring
+    # those errors.
+
+    @property
+    @abstractmethod
+    def value(self):
+        """The Python value"""
+        pass
+
+    @value.setter
+    @abstractmethod
+    def value(self, val):
+        pass
+
+    def __str__(self) -> str:
+        """
+        Produce a string representation for the instantiated type's Python value.
+        """
+        return str(self._value)  # type: ignore[attr-defined]
+
+    def clone(self):
+        """Clone this object. Structure is deepcopied, but values are not."""
+        return self.__class__()
```

Comment:
Ok, got it. Then I found this name and comment misleading. Maybe use `obj.__class__(...)` directly instead of this method? That would make it clear. Or maybe use different name. I also think that class method would be better., but it is pretty much just different syntax for the same thing. E.g.

```
@classmethod
def Create(cls, *args, **kwargs):
    """
    Create new instance of this class.
    """
    return cls(*args, **kwargs)
```